Fix: #75. Considerations for Intermediaries and Cache. (#76)
ioggstream/draft-polli-ratelimit-headers: draft-polli-ratelimit-headers.md

Under certain conditions, a server MAY artificially lower `RateLimit` field values between subsequent requests,
eg. to respond to Denial of Service attacks or in case of resource saturation.

# Intermediaries

An intermediary MUST NOT alter the RateLimit headers values
in such a way as to communicate a more permissive quota-policy;
this includes removing the RateLimit headers.

An intermediary MAY alter the RateLimit headers
in such a way as to communicate a more restrictive quota-policy when:

> - it is aware of the quota-unit semantic used by the Origin Server;

Author: ```suggestion
- it is aware of the quota-unit semantic used by the origin server;
```

- it implements this specification and enforces a quota-policy which
  is more restrictive than the one conveyed in the headers.

> An intermediary that does not implement this specification

Author: ```suggestion
An intermediary
```

@dret thanks, is that ok?

SHOULD forward a request even if it presumes that that request
might not be serviced;
the server returning the RateLimit headers is the sole responsible

> of enforcing the communicated quota-policy.

Reviewer: I understand this as meaning that the intermediary must not enforce rate limits that it did not further constrain, but it might be interpreted as if the enforcement should be happening at the same place where the headers are generated, which I think is not what you mean (although presumably such server knows some machine close to it would do the enforcing). Perhaps we could change a bit the language to remove ambiguities?

Author: Sure, feel free to propose an alternative text.

Author: @unleashed PTAL

Author: Hi @unleashed! I agree with you: can you propose a better text?

Reviewer: I think just replacing server with service might be enough. How about:

> An intermediary SHOULD forward a request even if it presumes that such request might not be serviced; the responsibility of enforcing the communicated quota-policy lies solely on the service that generated it.

That said, if the intermediary _itself_ needed to put some rate limiting in place (ie. to lower number of requests processed), there might be a good reason to allow restricting the quota and having the intermediary enforce it. Perhaps good to mention it as a known case in which a request would potentially not be forwarded?

Author: Agree: maybe we could clarify the issues without mandating strict behaviors.

This specification does not mandate any behavior on intermediaries
respect to retries,
nor requires that intermediaries have any role in respecting quota-policies.
For example, it is legitimate for a proxy to retransmit a request

> without notifying the client, and thus consuming quota-units.

Reviewer: I can see this being upsetting to clients, and didn't think through it before. After all, it's in the best interest of the clients to know if such intermediaries could consume resources. Which makes me question why would an intermediary perform a retry at the HTTP level that would end up consuming quota units (excluding intermediaries with some business logic around such resources).

Regardless, perhaps such intermediaries could notify clients of this possibility (or rather, that they will never do this) as a courtesy to help them manage expectations.

Edit: I see you add some text later on to suggest signalling the presence of this intermediary to the client. Assuming this would end up in this shape, maybe that should be encouraged a bit more strongly (if not outright mandated by the spec)?

Author: Yes, it's interest of user agents to know whether proxies consume or not the request.

> why would an intermediary perform a retry at the HTTP level that would end up consuming quota units?

I think we should not introduce constraint on infrastructure deployments, but only establish a semantic/syntax for the headers. I don't know why and if an intermediary can retry (maybe we can check http messaging spec), but if that's allowed, I'd avoid to interfere.

Reviewer: Fair enough. I expect in such cases the intermediaries that would be doing this to care a lot more about these headers than the clients, given the amount of control they would have.

# Caching

> The RateLimit headers MUST NOT be cached because their values are ephemeral.

Reviewer 2: These are going to get cached.  We probably need some other way to identify stale values.

Author: Ok, suggestions welcome! Preliminary considerations:

- if the response is served from a cache, then all subsequent requests using cached ratelimit values will hit the cache;
- the first request hitting the origin server, will retrieve and update the ratelimit values in the cache.
Does it sound reasonable?
 

WRT intermediaries: is that part ok?

Reviewer: I don't see how to proceed without adding some smarts to caches about these limits. If a cache is going to keep responding with stale data, then these headers are likely to matter a lot more to it than to the client originating the requests.

Author: @unleashed we should find a solution that avoids modifying caches.

If a user-agent uses cached ratelimit headers, at worst it will be throttled out (which is the same behavior that you'll get without the headers). In all the other cases, it will land on the cache.

Do you see any other problem?

If the user-agent feels it appropriate, it can use `Date` and `RateLimit-Reset` to check if the reset moment has passed or not.

Author: @unleashed PTAL

Reviewer: Yes, I was thinking about a way to relate the headers to a specific moment in time to let the user agent figure that the information is stale.

Given that having intermediaries deal with these headers is unlikely to happen anytime soon and the fact caches are going to keep passing the headers around, we could just add a provision to disregard the headers in case the user agent notices the information is stale.

The problem I see with the concept of stale information is how to determine it is actually stale. Using the `RateLimit-Reset` header has the problem that its value is relative and we allow an origin to return completely different data the next time a request is performed, even if it is within the described time window, so in some pathological cases a user agent might have trouble deducing whether these headers are stale or not.

Author: > how to determine it is actually stale

imho `Date` is the key. Both with a relative or an absolute timestamp, it's only via `Date` (or other response-timestamp field) that you can determine when the value was generated.

I agree that in patological use-cases clients should rely on the `Limit` and `window`, but this still happens with `retry-after`.

Reviewer: The change suggested looks fine. However I can see implementations trying to find out an algorithm to identify stale data once they become aware of the role intermediaries can play. Should we try to prescribe or hint at solutions in the text?

Author: We can provide guidance and ask the httpapi wg if those parts are needed or not. Probably warning implementors and let them decide is a reasonable way to do it.

Author: Avoid normative statements

```suggestion
```

Under certain conditions, a server MAY artificially lower `RateLimit` field values between subsequent requests,
eg. to respond to Denial of Service attacks or in case of resource saturation.

# Intermediaries

> An intermediary MUST NOT alter the RateLimit headers values

Author: ```suggestion

This section documents the considerations advised in Section 15.3.3 of {{SEMANTICS}}.

An intermediary MUST NOT alter the RateLimit headers values
```

in such a way as to communicate a more permissive quota-policy;
this includes removing the RateLimit headers.

An intermediary MAY alter the RateLimit headers
in such a way as to communicate a more restrictive quota-policy when:

- it is aware of the quota-unit semantic used by the Origin Server;
- it implements this specification and enforces a quota-policy which
  is more restrictive than the one conveyed in the headers.

An intermediary
SHOULD forward a request even if it presumes that that request
might not be serviced;

> the server returning the RateLimit headers is the sole responsible

Author: ```suggestion
the service returning the RateLimit headers is the sole responsible
```

of enforcing the communicated quota-policy.

This specification does not mandate any behavior on intermediaries
respect to retries,
nor requires that intermediaries have any role in respecting quota-policies.
For example, it is legitimate for a proxy to retransmit a request
without notifying the client, and thus consuming quota-units.

# Caching

The RateLimit headers MUST NOT be cached because their values are ephemeral.

> Apart from that, as is the ordinary case for HTTP caching ({{?RFC7234}}), a response with
> RateLimit header fields might be cached and re-used for subsequent requests.
> 

Author: ```suggestion
As is the ordinary case for HTTP caching ({{?RFC7234}}), a response with
RateLimit header fields might be cached and re-used for subsequent requests.
A cached RateLimit response, does not modify quota counters but could contain stale information.
Clients interested in determining the freshness of the RateLimit headers could rely on fields such as `Date` and on the `window` value of a `quota-policy`.
```

Reviewer: LGTM

Under certain conditions, a server MAY artificially lower `RateLimit` field values between subsequent requests,
eg. to respond to Denial of Service attacks or in case of resource saturation.

# Intermediaries

> 

Author: ```suggestion
```
